feat: calculate nRFCloud costs

Lenakh97 opened the pull request:
Closes #187

Lenakh97 committed:
feat: function and test for calculating nRFCloud costs

Lenakh97 committed:
feat: draft

Lenakh97 committed:
fix: add mock to test function

Lenakh97 committed:
feat: add api call to functions and refactor

Lenakh97 committed:
fix: change import

Lenakh97 committed:
fix: change import

Lenakh97 committed:
fix: add testdata to file

Lenakh97 committed:
refactor: move files and put test data into test file

Lenakh97 committed:
fix: add policies to lambda function for ssm and add test data to kpi

Lenakh97 committed:
fix: track costs for all accounts

Lenakh97 committed:
fix: move function into utils folder

Lenakh97 committed:
fix: add date to function parameters

Lenakh97 committed:
fix: increase timeout to 10 seconds

Lenakh97 committed:
fix: make sure calculateCosts can be tested with test data from august

Lenakh97 committed:
fix: change content type to application/json

coderbyheart commented on lambda/kpis.ts:
```diff
@@ -20,15 +24,27 @@ const { track, metrics } = metricsForComponent('KPIs')
+		costsPerAccount,
 	})
+	for (const acc in costsPerAccount) {
+		track(`nrfCloudPrices:${acc}`, MetricUnits.Count, costsPerAccount[acc] ?? 0)
```
```suggestion
		track(`nrfCloudMonthlyCosts:${acc}`, MetricUnits.Count, costsPerAccount[acc] ?? 0)
```

coderbyheart commented on nrfcloud/apiClient.ts:
```diff
@@ -206,6 +213,16 @@ export const apiClient = ({
+					'Content-Type': 'application/json',
+				},
+			})
+				.then<UsageSummary>(async (res) => res.json())
```
**note to self:** we should validate API responses

coderbyheart commented on nrfcloud/calculateCosts.ts:
```diff
@@ -0,0 +1,92 @@
+	WIFI: 0.002,
+}
+
+export const calculateCosts = async (
```
```suggestion
export const calculateCosts = (
```

This function does not need to be async.

coderbyheart commented on nrfcloud/calculateCosts.ts:
```diff
@@ -0,0 +1,92 @@
+	data: UsageSummary,
+	date: number,
+): Promise<number> => {
+	//return 0 if empty object
+	if (Object.keys(data).length === 0) {
+		return 0
+	}
```
```suggestion
```

Do not validate, what has been typed. The argument does not allow `{}` as data.

coderbyheart commented on nrfcloud/calculateCosts.ts:
```diff
@@ -0,0 +1,92 @@
+
+export type ServicePrice = keyof Prices
+
+export const nRFCloudPrices: Prices = {
```
Add link to source for prices.

coderbyheart commented on nrfcloud/calculateCosts.spec.ts:
```diff
@@ -0,0 +1,168 @@
+export const testDateAugust = 1691145383000
+
+describe('CalculateCosts()', () => {
+	it('should not calculate costs if given input is invalid.', async () => {
+		const expected = 0
+		expect(await calculateCosts({} as UsageSummary, testDateAugust)).toBe(
+			expected,
+		)
+	})
```
```suggestion
```

Not allowed as per API.

coderbyheart commented on nrfcloud/calculateCostsPerAccount.spec.ts:
```diff
@@ -0,0 +1,93 @@
+import { calculateCostsPerAccount } from './calculateCostsPerAccount.js'
+import { testDateAugust } from './calculateCosts.spec.js'
+
+jest.mock('.././nrfcloud/allAccounts.js', () => ({
+	getAllnRFCloudAccounts: async () => {
+		return ['account1', 'account2']
+	},
+}))
```
Avoid global mocking, instead use dependency injection (see below).

coderbyheart commented on nrfcloud/calculateCostsPerAccount.spec.ts:
```diff
@@ -0,0 +1,93 @@
+	it('should calculate costs per Account', async () => {
+		const expected = { account1: 40.15, account2: 40.15 }
+		expect(
+			await calculateCostsPerAccount({
+				ssm,
+				stackName: STACK_NAME,
+				date: testDateAugust,
+			}),
```
```suggestion
			await calculateCostsPerAccount({
				ssm,
				stackName: STACK_NAME,
				date: testDateAugust,
				{getAllnRFCloudAccounts: () => ['account1', 'account2']}
			}),
```

Pass in collaborators, this makes it clearer what the unit under test uses (because you it is not obvious that the mock here is actually used).

coderbyheart commented on nrfcloud/calculateCostsPerAccount.ts:
```diff
@@ -0,0 +1,27 @@
+
+export type AccountObject = { [key: string]: number }
+
+export const calculateCostsPerAccount = async ({
+	ssm,
+	stackName,
+	date,
+}: {
+	ssm: SSMClient
+	stackName: string
+	date: number
+}): Promise<AccountObject> => {
```
Change the lambda function to not use the global `getAllnRFCloudAccounts()` but the function passed as the argument.

coderbyheart commented on util/getCurrentMonth.spec.ts:
```diff
@@ -0,0 +1,8 @@
+import { testDateAugust } from '../nrfcloud/calculateCosts.spec'
```
Don't re-use simple constants (so the other test can change it without affecting this test).

coderbyheart commented on util/getCurrentMonth.ts:
```diff
@@ -0,0 +1,6 @@
+export const getCurrentMonth = (date: Date): string => {
+	const currentDate = `${date.getFullYear()}-${(date.getMonth() + 1)
+		.toString()
+		.padStart(2, '0')}`
+	return currentDate
+}
```
```suggestion
export const getCurrentMonth = (date: Date): string => date.toISOString().slice(0, 7)

coderbyheart reviewed (changes requested):
Looking good!

Lenakh97 committed:
Update lambda/kpis.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
Update util/getCurrentMonth.ts

Co-authored-by: Markus Tacker <Markus.Tacker@NordicSemi.no>

Lenakh97 committed:
fix: remove async function and remove validation
Co-authored-by: Markus Tacker  markus.tacker@nordicsemi.no

Lenakh97 committed:
fix: make testDateAugust a local variable

Lenakh97 committed:
fix: make testDateAugust a local variable

Lenakh97 committed:
fix: pass functions as argument

Lenakh97 committed:
fix: add missing name

coderbyheart commented on nrfcloud/calculateCosts.ts:
```diff
@@ -0,0 +1,86 @@
+	WIFI: 0.002,
+}
+
+export const calculateCosts = (data: UsageSummary, date: number): number => {
```
```suggestion
export const calculateCosts = (data: UsageSummary, date: Date): number => {
```

Always use `Date` type.

coderbyheart commented on nrfcloud/calculateCosts.ts:
```diff
@@ -0,0 +1,86 @@
+			}
+		}
+	}
+	monthlyCosts = Number((monthlyCosts > 1.99 ? monthlyCosts : 1.99).toFixed(2))
```
This is a unnecessary conversion of a number to a string through `toFixed`. Never represent numbers as strings when doing Math.
Especially when it comes to monetary calculations. You are cutting of fractions of a cent. While here this is not significant, this small mistake, if repeatedly applied, can result in big errors.
Let the end-use decide how to format fractions.

coderbyheart replied:
If you want to avoid fractions, compute everything in cents.

coderbyheart commented on nrfcloud/calculateCostsPerAccount.spec.ts:
```diff
@@ -0,0 +1,121 @@
+import { STACK_NAME } from '../cdk/stacks/stackConfig.js'
+import { SSMClient } from '@aws-sdk/client-ssm'
+import { calculateCostsPerAccount } from './calculateCostsPerAccount.js'
+const testDateAugust = 1691145383000
```
```suggestion
const testDateAugust = new Date('2023-08-04T10:36:23.000Z')
```
is much more readable.

coderbyheart commented on nrfcloud/getCostSummaryFromAPI.ts:
```diff
@@ -0,0 +1,28 @@
+
+	if ('error' in summary) {
+		console.error(summary.error)
+		process.exit(1)
```
**Problem:** NEVER terminate the process from within a utility function. This is only useful for CLI scripts, but especially not in Lambdas.

coderbyheart replied:
Instead, return an error.

coderbyheart committed:
refactor: use Date, remove intermediate function

Use Date when passing along a date.
Removed two intermediate functions which can be handled in the lambda directly.
Do not process.exit() on error
Removed truncation of calculated price
